### PR TITLE
Detached Root Data Store creation

### PR DESCRIPTION
**Pull request by vladsud**

### CODE REVIEW

File: packages/runtime/container-runtime-definitions/src/containerRuntime.ts [hunk 1]
```diff
@@ -109,7 +110,8 @@ export interface IContainerRuntime extends
     createRootDataStore(pkg: string | string[], rootDataStoreId: string): Promise<IFluidRouter>;
 
     /**
-     * Returns the current quorum.
+     * Creates detached data store context. only after context.attachRuntime() is called,
```
Comment (heliocliu) [inline packages/runtime/container-runtime-definitions/src/containerRuntime.ts:113]: revert this change

File: packages/runtime/container-runtime-definitions/src/containerRuntime.ts [hunk 2]
```diff
@@ -118,6 +120,14 @@ export interface IContainerRuntime extends
      */
     getAudience(): IAudience;
 
+    /**
+     * Creates detached data store context. only after context.attachRuntime() is called,
+     * data store initialization is considered compete.
```
Comment (heliocliu) [inline packages/runtime/container-runtime-definitions/src/containerRuntime.ts:125]:
```suggestion
     * Creates detached data store context. Data store initialization is considered compete
     * only after context.attachRuntime() is called.
```

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 3]
```diff
@@ -1361,18 +1361,33 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         return fluidDataStore;
     }
 
-    public createDetachedDataStore(): IFluidDataStoreContextDetached {
-        const id = uuid();
+    public createRootDetachedDataStore(
```
Comment (heliocliu) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1364]: Feels a bit pedantic and hopefully another native English speaker can comment as well in case I'm wrong, but...

I think this should be `createDetachedRootDataStore`?  Partly because it's a root data store that is detached (as opposed to a detached data store that is root), and partly because (I _think_) it violates the unspoken rule of adjective ordering?

File: packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts [hunk 4]
```diff
@@ -292,25 +289,53 @@ export class PureDataObjectFactory<TObj extends PureDataObject<O, S, E>, O, S, E
...
+            runtime,
+            [this.type],
+            initialState);
     }
 
-    protected async createInstanceCore(
+    /**
+     * Creates a new root instance of the object. Uses container's registry to find this factory.
+     * It's expected that only container owners would use this functionality, as only such developers
+     * have knowledge of entries in container registry.
```
Comment (heliocliu) [inline packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts:301]: It feels just a bit weird to have this comment (and I guess the `IContainerRuntime` typing below) playing primary gatekeeper to using this API when it lets you specify the data store id whereas the alternate API does not.  Hopefully no one abuses it 🙃

**Comment by heliocliu** [comment]
Can you add a brief description on the motivation for this change?  So anyone who chances upon the PR/commit doesn't need to crossreference for missing context 🙂

**Review by heliocliu — APPROVED** [review]

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 3 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1364]: I agree :)

File: packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts (hunk 4 above)
Comment (vladsud) [reply packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts:301]: I think this is the right approach: https://github.com/microsoft/FluidFramework/pull/3700
But while we wait to settle it, I'd use any other opportunity to protect API surface against misuse.
Yes, it's easy to overcome it (simple cast), but it's at least something.
